### PR TITLE
Keep volume policies after migrating it to another primary storage

### PR DESCRIPTION
**Pull request by GutoVeronezi**
### Description
In ACS, there are two ways to migrate a volume between storages:
- migrating only the volume (`migrateVolume`);
- migrating a VM with volume (`migrateVirtualMachineWithVolume`);

Both, although are in differents flows, work similarly: they create a new volume in the destination storage and then delete the old volume in the source storage; When we delete the old volume, we also delete the volume policies (snapshot scheduling). 

Therefore, the new volume will not have the snapshot schedules and we will lose this information. This PR intends to copy the policies from source to destination volume before deleting the old volume;

### Types of changes
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale
- [ ] Major
- [x] Minor

### How Has This Been Tested?
It has been tested locally in a test lab.
1. I created a VM;
2. I configured snapshot schedules in the volume;
3. I migrated :
   - volume (with `migrateVolume`);
   - VM and volume (with `migrateVirtualMachineWithVolume`);
4. I checked if the policies were copied to new volume;   

-- After applying this patch, the policies were copied.
Also, I created Java unit tests for the methods I changed.

### CODE REVIEW

File: engine/storage/volume/src/main/java/org/apache/cloudstack/storage/volume/VolumeServiceImpl.java [hunk 1]
```diff
@@ -1868,6 +1847,65 @@ protected Void copyVolumeCallBack(AsyncCallbackDispatcher<VolumeServiceImpl, Cop
         return null;
     }
 
+    @Override
+    public boolean copyPoliciesBetweenVolumesAndDestroySourceVolumeAfterMigration(Event destinationEvent, Answer destinationEventAnswer, VolumeInfo sourceVolume,
+      VolumeInfo destinationVolume, boolean retryExpungeVolumeAsync) {
```
Comment (sureshanaparti) [inline engine/storage/volume/src/main/java/org/apache/cloudstack/storage/volume/VolumeServiceImpl.java:1852]: @GutoVeronezi I think, it is better to separate copy policies and destroy volume.

Comment (GutoVeronezi) [reply engine/storage/volume/src/main/java/org/apache/cloudstack/storage/volume/VolumeServiceImpl.java:1852]: @sureshanaparti I extracted the code that destroy the source volume to a method and added tests to it; However I kept this method, calling both `copy policies` and `destroy volume` to facilitate and guide the process.

**Comment by GutoVeronezi** [comment]
Can anyone review this?

**Review by DaanHoogland — APPROVED** [review]
This is a lot of good improvements in log and exception messages as well as a solid looking re-factor of the volume object and - service. I t will need extensive regression testing on anything that relates to that. I don;t see hypervisor specific risks but these may be hidden in there.
CLGTM

File: engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/VolumeService.java [hunk 2]
```diff
@@ -100,4 +101,16 @@ public VolumeInfo getVolume() {
...
 
     void unmanageVolume(long volumeId);
+
+    /**
+     * After volume migration, copies snapshot policies from the source volume to destination volume; then, it destroys and expunges the source volume.
+     * @param destinationEvent
+     * @param destinationEventAnswer
+     * @param sourceVolume
+     * @param destinationVolume
+     * @param retryExpungeVolumeAsync
```
[the remark this answers is not in the document]
Comment (GutoVeronezi) [reply engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/VolumeService.java]: @DaanHoogland removed, thanks.

File: engine/storage/volume/src/test/java/org/apache/cloudstack/storage/volume/VolumeObjectTest.java [hunk 3]
```diff
@@ -19,59 +19,558 @@
...
+
+        volumeObjectSpy.handleProcessEventAnswer(createObjectAnswer);
+
+        Mockito.verify(volumeObjectSpy, Mockito.times(1)).handleProcessEventAnswer(createObjectAnswer, true);
     }
 
-    /**
-     * Tests the following scenario:
-     * If the volume gets deleted by another thread (cleanup) and the cleanup is attempted again, the volume isnt found in DB and hence NPE occurs
-     * during transition
```
[the remark this answers is not in the document]
Comment (GutoVeronezi) [reply engine/storage/volume/src/test/java/org/apache/cloudstack/storage/volume/VolumeObjectTest.java:70]: @DaanHoogland This whole test class, as well as those in test package in this module were deleted. The tests in this module were already disabled for a long time; therefore we considered that they were irrelevant, as they were a bunch of code that were not being used.

[the remark this answers is not in the document]
Comment (DaanHoogland) [reply engine/storage/volume/src/test/java/org/apache/cloudstack/storage/volume/VolumeObjectTest.java:70]: ah right, it isn't a change but an addition and a deletion, sorry and thanks

**Comment by GutoVeronezi** [comment]
Can anyone review this?

**Comment by DaanHoogland** [comment]
@GutoVeronezi a trick for getting the right person is running `git blame` on the files and see if there is the name of an active contributor there to ask directly.

**Comment by GutoVeronezi** [comment]
> @GutoVeronezi a trick for getting the right person is running `git blame` on the files and see if there is the name of an active contributor there to ask directly.

Thanks for the tip @DaanHoogland.

**Comment by DaanHoogland** [comment]
@sureshanaparti does this LGTY? cc @nvazquez

**Comment by nvazquez** [comment]
Looking good, I think it will need manual testing for migrations and also volumes regression testing, @sureshanaparti is this something you can check please?

**Comment by GutoVeronezi** [comment]
@nvazquez any update about this one?

**Comment by nvazquez** [comment]
@weizhouapache can you please review/verify this improvement?

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Review by weizhouapache — APPROVED** [review]
tested ok on kvm.

(1) both root and data disk
(2) offline volume migration (migratevolume)
(3) online volume migration (migratevirtualmachinewithvolume)

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

**Review by nvazquez — APPROVED** [review]
LGTM

**Pull request merged**